torch/csrc/api/include/torch/nn/functional/loss.h

```diff
@@ -383,9 +383,26 @@ inline Tensor smooth_l1_loss(
 inline Tensor smooth_l1_loss(
     const Tensor& input,
     const Tensor& target,
-    const SmoothL1LossFuncOptions& options = {},
-    double beta = 1.) {
-  return detail::smooth_l1_loss(input, target, options.reduction(), beta);
+    const SmoothL1LossFuncOptions& options = {}) {
+  return detail::smooth_l1_loss(
+      input, target, options.reduction(), options.beta());
+}
+
+/// See
+/// https://pytorch.org/docs/master/nn.functional.html#torch.nn.functional.smooth_l1_loss
+/// about the exact behavior of this functional.
+///
+/// Example:
+/// ```
+/// namespace F = torch::nn::functional;
+/// F::smooth_l1_loss(input, target, /*reduction=*/torch::kNone, /*beta=*/0.5);
+/// ```
+inline Tensor smooth_l1_loss(
+    const Tensor& input,
+    const Tensor& target,
+    SmoothL1LossFuncOptions::reduction_t reduction,
```

Comment:
actually, should this be just `const SmoothL1LossFuncOptions& options = {}` to keep BC?

Reply:
`smooth_l1_loss(..., const SmoothL1LossFuncOptions& options = {}, double beta)`  doesn't build because a default argument cannot be followed by a non-default argument.

`smooth_l1_loss(..., const SmoothL1LossFuncOptions& options, double beta)`  does build, but it's semantically ambiguous because beta is given in two places. For instance, what should happen if it's called like `smooth_l1_loss(input, target, SmoothL1LossFuncOptions().beta(0.5), /*beta=*/0.1)`?

With the signatures in this PR, it's no longer possible to call the function in an ambiguous way. In that sense, it is BC-breaking. However, I would think that BC does not actually matter for ambiguous cases. Maybe I'm mistaken though

We do keep BC for the non-ambiguous case `smooth_l1_loss(input, target, /*reduction=*/..., /*beta=*/...)`. There is an existing test for this in `test/cpp/api/functional.cpp`

Reply:
Although if we do break BC to remove the ambiguous case, maybe we would still need a deprecation cycle. `torch_extensions` builds in CI are failing from my change:

```
2023-09-21T18:54:43.9035006Z /var/lib/jenkins/.cache/torch_extensions/py311_cpu/functional_impl_check/main.cpp:1900:21: error: no matching function for call to 'smooth_l1_loss'
2023-09-21T18:54:43.9035367Z   auto cpp_output = F::smooth_l1_loss(
2023-09-21T18:54:43.9035584Z                     ^~~~~~~~~~~~~~~~~
2023-09-21T18:54:43.9036468Z /opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/include/torch/csrc/api/include/torch/nn/functional/loss.h:400:15: note: candidate function not viable: no known conversion from 'torch::nn::SmoothL1LossOptions' to 'SmoothL1LossFuncOptions::reduction_t' (aka 'variant<enumtype::kNone, enumtype::kMean, enumtype::kSum>') for 3rd argument
2023-09-21T18:54:43.9037033Z inline Tensor smooth_l1_loss(
```

Perhaps another strategy is to go with `smooth_l1_loss(..., const SmoothL1LossFuncOptions& options, double beta)` after all, keeping the ambiguous case, but error out if `beta` is given in `options`. In order to do that though, we would need to change the `beta` field in `SmoothL1LossFuncOptions` from `double` to `optional<double>`, so we can detect if it was explicitly set or not. Is this alright with you?

This would still be BC-breaking in cases where both `beta` parameters are set, like `smooth_l1_loss(input, target, SmoothL1LossFuncOptions().beta(0.5), /*beta=*/0.1)`, but at least that one case has to be broken in order to address the issue

Reply:
> Perhaps a better strategy is to go with `smooth_l1_loss(..., const SmoothL1LossFuncOptions& options, double beta)` after all, but just error out in ambiguous case where both `beta` parameters are set like `smooth_l1_loss(input, target, SmoothL1LossFuncOptions().beta(0.5), /*beta=*/0.1)`. This should make the failing builds pass again. In order to do this though, we would need to change the `beta` field in `SmoothL1LossFuncOptions` from `double` to `optional<double>`, so we can detect if it was explicitly set or not.

I've changed the PR to do this

Reply:
@mikaylagawarecki, are these changes alright with you?